hypothesis-python/src/hypothesis/extra/numpy.py

```diff
@@ -374,15 +375,18 @@ def fill_for(elements, unique, fill, name=""):
     return fill
 
 
+D = TypeVar("D", bound=DTypeLike)
+
+
 @defines_strategy(force_reusable_values=True)
 def arrays(
-    dtype: Any,
+    dtype: D,
     shape: Union[int, st.SearchStrategy[int], Shape, st.SearchStrategy[Shape]],
     *,
     elements: Optional[Union[st.SearchStrategy, Mapping[str, Any]]] = None,
     fill: Optional[st.SearchStrategy[Any]] = None,
```

Comment:
I don't think there's any nice way to constrain these strategies to compatibility with `D`?

Reply:
Don `dtype`, `elements`, `fill` and `returned searchstrategy`  all have the same dtype or dtype only can say something about the `returned searchstrategy`.

If `elements`, `fill` must be of the dtype `dtype` then we could use D for these two cases too.

Reply:
They can be different, so long as casting is lossless. So probably best to leave it as-is for now.